File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
    private void swap(int[] v, int a, int b) {
        v[a] = v[a] ^ v[b];
        v[b] = v[a] ^ v[b];
        v[a] = v[a] ^ v[b];
    }

    public int firstMissingPositive(int[] v) {
        int n = v.length;
        for (int i = 0; i < n; i++) {
            while (v[i] > 0 && v[i] <= n && v[v[i] - 1] != v[i]) {
                swap(v, i, v[i] - 1);
            }
        }
        for (int i = 0; i < n; i++) {
            if (v[i] != i + 1) {
                return i + 1;
            }
        }
        return n + 1;
    }
}

File: 0041-first-missing-positive/0041-first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& v) {
        int n = v.size();
        vector<int> mp(n + 1, 0);

        for (auto e : v) {
            if (e > 0 && e <= n) {
                mp[e]++;
            }
        }

        for (int i = 1; i <= n; i++) {
            if (mp[i] == 0) {
                return i;
            }
        }

        return n + 1;
    }
};
